src/aof.c

```diff
@@ -1184,99 +1615,174 @@ void backgroundRewriteDoneHandler(int exitcode, int bysignal) {
-         * caused by rename(2) in scenario 1, or by close(2) in scenario 2), we
-         * use a background thread to take care of this. First, we
-         * make scenario 1 identical to scenario 2 by opening the target file
-         * when it exists. The unlink operation after the rename(2) will then
-         * be executed upon calling close(2) for its descriptor. Everything to
-         * guarantee atomicity for this switch has already happened by then, so
-         * we don't care what the outcome or duration of that close operation
-         * is, as long as the file descriptor is released again. */
-        if (server.aof_fd == -1) {
-            /* AOF disabled */
+#ifdef USE_NVML
+        if (server.aof_use_nvml == 1) {
+            /*size_t real_log_rewr_size; */
+            if (server.aof_nvml_direct == 1) {
+                pmemlog_walk(server.aof_plp, 0, copyDiffPMEMlog, NULL);
+            } else {
+                if (aofRewriteBufferWritePMEMlog(server.aof_plp_rewr) == -1) {
+                    redisLog(REDIS_WARNING,
+                        "Error trying to flush the parent diff to the rewritten AOF: %s", strerror(errno));
+                    goto cleanup;
+                }
+            }
+            latencyEndMonitor(latency);
+            latencyAddSampleIfNeeded("aof-rewrite-diff-write",latency);
+
+            redisLog(REDIS_NOTICE,
+                "Parent diff successfully flushed to the rewritten AOF (%lu bytes)", aofRewriteBufferSize());
+
+            /* swap filenames, swap fd's  and plp's */
+            latencyStartMonitor(latency);
+            snprintf(tmpfile, 256, "aof-rewrite-swap-%d.aof", getpid());
+            if (rename (server.aof_filename, tmpfile) == -1) {
+                redisLog(REDIS_WARNING, "Redis cannot rename AOF to temporary file: %s",
+                        strerror(errno));
+                goto cleanup;
+            }
+            if (rename(server.aof_filename_rewr, server.aof_filename) == -1) {
+                redisLog(REDIS_WARNING, "Redis cannot rename AOF rewrite file to AOF: %s",
+                        strerror(errno));
+                goto cleanup;
+            }
+            if (rename(tmpfile, server.aof_filename_rewr) == -1) {
+                redisLog(REDIS_WARNING, "Redis cannot rename temporary file to AOF rewrite: %s",
+                        strerror(errno));
+                goto cleanup;
+            }
+            latencyEndMonitor(latency);
+            latencyAddSampleIfNeeded("aof-rename",latency);
+
+            if (server.aof_state != REDIS_AOF_OFF) {
+
+                PMEMlogpool *tmp_plp = server.aof_plp;
+                server.aof_plp = server.aof_plp_rewr;
+                server.aof_plp_rewr = tmp_plp;
+
+                /* compute real size of rewrite file, if it's different, than configured
+                 * the reopen file */
+                if ((pmemlog_tell(server.aof_plp) + pmemlog_nbyte(server.aof_plp)) !=
```

Comment:
1) You're right, pmemlog_tell() shouldn't be here.
2) Because configured pool size is different than usable space after pool creation.